firebase_functions/functions/src/helpers.ts

```diff
@@ -163,11 +163,60 @@ const faucet = async receiveAddress => {
   );
 };
 
+
+const getBalance = async address => {
+  try {
+    if (!address) {
+      return 0;
+    }
+    const { provider } = await getSettings();
+    const { getBalances } = composeAPI({ provider });
+    const { balances } = await getBalances([address], 100);
+    return balances && balances.length > 0 ? balances[0] : 0;
+  } catch (error) {
+    console.error('getBalance error', error);
+    return 0;
+  }
+};
+
+
+const repairWallet = async (seed, keyIndex) => {
+  try {
+    //Iterating through keyIndex ordered by likelyhood
+    for (let value of [-2, -1, 1, 2, 3, 4, -3, -4, -5, -6, -7, 5, 6, 7]) {
+      const newIndex = Number(keyIndex) + Number(value)
+      if (newIndex >= 0) {
+        const newAddress = await generateAddress(seed, newIndex)
+        const newBalance = await getBalance(newAddress);
+        if (newBalance > 0) {
+          return { address: newAddress, keyIndex: newIndex };
+        }
+      }
+    }
+  } catch (error) {
+    console.log("Repair wallet Error", error)
+    return error;
```

Comment:
The edge case still remains here.
if you return an error, and not the object containing `{address, keyIndex}`,
then the call to the function will result in an exception, once you try to access newIotaWallet.address
```
const newIotaWallet :any = await repairWallet(seed, keyIndex)
   address = newIotaWallet.address;
   keyIndex= newIotaWallet.keyIndex;
```

Also, for the successful case, there is no default return object containing `{address, keyIndex}`. 
If in a vey unlikely but possible case, none of the values will match the `if (newIndex >= 0)` condition, the function will end without returning anything. so Call to `const newIotaWallet :any = await repairWallet(seed, keyIndex); 
address = newIotaWallet.address;` will produce an exception

Reply:
Makes sense. But what do we want to happen? 
Should we return the previous incorrect {seed, keyIndex} pair?

Reply:
We still can return an error or null. But on the receiving side, where we expect the object to be returned, we can add a check to see if the required fields exists.
In a very simple case: 
```
if (newIotaWallet && newIotaWallet.address &&newIotaWallet.keyIndex) {
   address = newIotaWallet.address;
   keyIndex= newIotaWallet.keyIndex;
}
```